### DISCUSSION TITLE
Testing and feedback needed for upcoming ADMB-13.0 release

### DISCUSSION THREAD

**Discussion by johnoel** [admb-project/admb#249] [General]
Hi all,

ADMB-13 will be the next major release.  It is scheduled to be released at the near the end of May 2022.  To ensure a smooth release, user testing is requested.  Please download and build the [admb-13.0](https://github.com/admb-project/admb/tree/admb-13.0) branch and test with existing models.  The major release has many [changes](https://github.com/admb-project/admb/blob/admb-13.0/CHANGES.md) and new features included.  One of the big changes is improving the program display (see issue #221).  The display is more concise and provides additional diagnostics such as bounds checking.  @Cole-Monnahan-NOAA and @wStockhausen lead the development.  One major feature that was developed is to run the body of a for loop in parallel instead of sequentially.  The forest example was [modified](https://github.com/admb-project/admb/blob/admb-13.0/tests/thread_funnel/forest/forest.tpl) to run the funnel function in parallel.  The modified example has shown improved runtime performance compared to the original version.  So, please try out the release and provide feedback  and report any errors that were found.

Thank you,
Johnoel

**Comment by Cole-Monnahan-NOAA** [comment]
@johnoel At least two of us can't get it to run, see [this issue](https://github.com/nmfs-stock-synthesis/stock-synthesis/issues/257). Could you provide more details on how to install? Thanks!

**Reply by k-doering-NOAA** [reply]
Also having issues compiling locally with mingw64 on windows.

**Reply by johnoel** [reply]
Corrected issue.  Please pull updates and try again.

**Reply by k-doering-NOAA** [reply]
Hi Johnoel, I still wasn't able to compile locally after pulling in updates. Here is what I am seeing ( I think this was different than the errors others were encountering, though):

```
C:\Users\Kathryn.Doering\Documents\github_repos\admb>make g++-dist DEBUG=
make g++-core
make[1]: Entering directory '/c/Users/Kathryn.Doering/Documents/github_repos/admb'
make --directory=src CC=gcc CXX=g++ all
make[2]: Entering directory '/c/Users/Kathryn.Doering/Documents/github_repos/admb/src'
make _objects
make[3]: Entering directory '/c/Users/Kathryn.Doering/Documents/github_repos/admb/src'
g++ -c -std=c++14 -O2 -Wall -Wextra -Wconversion -Wno-unused-result -Wno-unused-parameter -DOPT_LIB -D_USE_MATH_DEFINES -I../build/admb/include -o ../build/objects-mingw64-g++8/optlp-nh99-model9.obj nh99/model9.cpp
nh99/model9.cpp: In function 'int operator+(int, const data_int&)':
nh99/model9.cpp:12:14: error: 'int data_int::val' is protected within this context
   return n+v.val;
              ^~~
In file included from nh99/model9.cpp:8:
../build/admb/include/admodel.h:1484:7: note: declared protected here
   int val;
       ^~~
nh99/model9.cpp: In function 'int operator+(const data_int&, int)':
nh99/model9.cpp:17:14: error: 'int data_int::val' is protected within this context
   return n+v.val;
              ^~~
In file included from nh99/model9.cpp:8:
../build/admb/include/admodel.h:1484:7: note: declared protected here
   int val;
       ^~~
nh99/model9.cpp: In function 'int operator+(const data_int&, const data_int&)':
nh99/model9.cpp:22:12: error: 'int data_int::val' is protected within this context
   return n.val + v.val;
            ^~~
In file included from nh99/model9.cpp:8:
../build/admb/include/admodel.h:1484:7: note: declared protected here
   int val;
       ^~~
nh99/model9.cpp:22:20: error: 'int data_int::val' is protected within this context
   return n.val + v.val;
                    ^~~
In file included from nh99/model9.cpp:8:
../build/admb/include/admodel.h:1484:7: note: declared protected here
   int val;
       ^~~
make[3]: *** [GNUmakefile:630: ../build/objects-mingw64-g++8/optlp-nh99-model9.obj] Error 1
make[3]: Leaving directory '/c/Users/Kathryn.Doering/Documents/github_repos/admb/src'
make[2]: *** [GNUmakefile:551: objects] Error 2
make[2]: Leaving directory '/c/Users/Kathryn.Doering/Documents/github_repos/admb/src'
make[1]: *** [Makefile:97: g++-core] Error 2
make[1]: Leaving directory '/c/Users/Kathryn.Doering/Documents/github_repos/admb'
make: *** [Makefile:92: g++-dist] Error 2

```

**Reply by johnoel** [reply]
Did you clean first?

```
> make clean
```

**Reply by k-doering-NOAA** [reply]
> Did you clean first?
> 
> ```
> > make clean
> ```

Nope, I did not! Thanks @johnoel . I compiled successfully now!

**Comment by Rick-Methot-NOAA** [comment]
successful compile for me too, but the installer reports that it cannot find many files in examples/admb-re  folders.
looking forward to giving it a try.

**Reply by johnoel** [reply]
are you using?

```
> make install 
```

Can you post the errors?

**Comment by Rick-Methot-NOAA** [comment]
errors looked like:
Could Not Find c:\ADMB-13.0\examples\admb-re\pheno\pheno.bar
Could Not Find c:\ADMB-13.0\examples\admb-re\pheno\pheno.bgs
Could Not Find c:\ADMB-13.0\examples\admb-re\pheno\pheno.cor
Could Not Find c:\ADMB-13.0\examples\admb-re\pheno\pheno.cpp
and many more files not found.

**Reply by johnoel** [reply]
What command did you use to get the errors?

**Reply by johnoel** [reply]
Was it?

```
> make clean
```

**Comment by Rick-Methot-NOAA** [comment]
make clean

**Reply by johnoel** [reply]
Those can be ignored.  Should be better though...

**Comment by Cole-Monnahan-NOAA** [comment]
I can also now compile and run locally. I found no errors or issues on my small test set. 

Windows 10. g++ (Built by Jeroen for the R-project) 8.3.0

**Comment by Rick-Methot-NOAA** [comment]
When trying to compile ss3 locally, I got this error:

*** Compile: ss.cpp
g++ -c -std=c++14 -O3 -fpermissive -D_FILE_OFFSET_BITS=64 -DUSE_ADMB_CONTRIBS -D_USE_MATH_DEFINES -I. -I"C:\ADMB-13.0\include" -I"C:\ADMB-13.0\include\contrib" -o ss.obj ss.cpp
ss.cpp: In member function 'void model_parameters::write_bigoutput()':
ss.cpp:29906:27: error: 'get_ln_det_value' was not declared in this scope
   temp = get_ln_det_value();
                           ^

**Reply by johnoel** [reply]
I can't seem to duplicate the compile error.  How did you build ss.tpl?

**Comment by Rick-Methot-NOAA** [comment]
I simply ran tpl2cpp, then ran admb

**Reply by johnoel** [reply]
Looks like you need to clean and recompile again.  I can send you a precompiled admb directory for testing.

```
C:\stock-synthesis-github\ss330>C:\Users\johnoel\admb-github\admb.cmd ss.tpl

*** Parse: ss.tpl
xxglobal.tmp
xxhtop.tmp
header.tmp
xxalloc.tmp
xxtopm.tmp
        1 file(s) copied.
tpl2cpp   ss

*** Compile: ss.cpp
g++ -c -std=c++14 -O2 -D_FILE_OFFSET_BITS=64 -DUSE_ADMB_CONTRIBS -D_USE_MATH_DEFINES -I. -I"C:\Users\johnoel\admb-github\build\admb\include" -I"C:\Users\johnoel\admb-github\build\admb\include\contrib" -o ss.obj ss.cpp

*** Linking: ss.obj
g++ -static -o ss.exe ss.obj "C:\Users\johnoel\admb-github\build\admb\lib\libadmb-contrib-mingw64-g++8.a"

Successfully built 'ss.exe'.
```

**Comment by wStockhausen** [comment]
And now for some good news: I had no problems on my Mac running OSX
Catalina (I hate to give up the desktop picture). Just compiled and ran my
Tanner crab assessment model--no issues!

*William T. Stockhausen*

Research Fishery Biologist, Alaska Fisheries Science Center

NOAA Fisheries | U.S. Department of Commerce

Office: (206) 526-4241

www.fisheries.noaa.gov

**Comment by Rick-Methot-NOAA** [comment]
switched to RTools and got a 
"Successfully built executable."
message from make

**Reply by johnoel** [reply]
Was it msys2 that did not work?

**Reply by Rick-Methot-NOAA** [reply]
Johnoel,
even though the installation reported being successful, it still did not set up path and environment such that I could compile using admb.cmd

**Reply by johnoel** [reply]
See Issue #260.  Should make it easier to set PATH for Windows.

**Comment by k-doering-NOAA** [comment]
@johnoel wanted to report that the ADMB 13 branch works well for Stock Synthesis models! Comparable runtimes, and no change in results for the tested models. We are also able to use ADMB 13 to compile SS3 on Mac, Linux, and Windows. We also like the new console, and it has inspired us to clean up some of the console output that SS3 adds on. Thank you for the option to test in advance before releasing!

**Reply by k-doering-NOAA** [reply]
Sorry, one more thing: these are the compiler warnings we get with -Wall . Just reporting this in case some of them should be fixed in admb: https://github.com/nmfs-stock-synthesis/stock-synthesis/issues/257#issuecomment-1128944492

**Reply by johnoel** [reply]
See Issue #251

**Reply by k-doering-NOAA** [reply]
Thanks for fixing those! I was just highlighting some other compiler warnings that we see when compiling stock synthesis with either ADMB 12.3 or 13, but perhaps they are safe to ignore (as we have been doing for some time)

**Comment by johnoel** [comment]
Sorry all, need to delay release of ADMB-13 till next month.  Working on Issue #255.
